source/common/router/config_impl.cc

```diff
@@ -1879,20 +1879,50 @@ RouteConstSharedPtr VirtualHostImpl::getRouteFromEntries(const RouteCallback& cb
       const auto result = match.result_();
       if (result->typeUrl() == RouteMatchAction::staticTypeUrl()) {
         const RouteMatchAction& route_action = result->getTyped<RouteMatchAction>();
+        if (!headers.Path() && !route_action.route()->supportsPathlessHeaders()) {
+          return nullptr;
+        }
 
-        if (route_action.route()->matches(headers, stream_info, random_value)) {
-          return route_action.route();
+        RouteConstSharedPtr route_entry =
+            route_action.route()->matches(headers, stream_info, random_value);
+        if (nullptr == route_entry) {
+          ENVOY_LOG(debug, "route was resolved but final route did not match incoming request");
+          return nullptr;
         }
 
-        ENVOY_LOG(debug, "route was resolved but final route did not match incoming request");
-        return nullptr;
+        if (cb) {
+          if (cb(route_entry, RouteEvalStatus::NoMoreRoutes) == RouteMatchStatus::Accept) {
+            return route_entry;
+          }
+          return nullptr;
+        }
+
+        return route_entry;
       } else if (result->typeUrl() == RouteListMatchAction::staticTypeUrl()) {
         const RouteListMatchAction& action = result->getTyped<RouteListMatchAction>();
 
-        for (const auto& route : action.routes()) {
-          if (route->matches(headers, stream_info, random_value)) {
-            return route;
+        for (auto route = action.routes().begin(); route != action.routes().end(); ++route) {
+          if (!headers.Path() && !(*route)->supportsPathlessHeaders()) {
+            continue;
           }
+
+          RouteConstSharedPtr route_entry = (*route)->matches(headers, stream_info, random_value);
+          if (nullptr == route_entry) {
+            continue;
+          }
+
+          if (cb) {
+            RouteEvalStatus eval_status = (std::next(route) == action.routes().end())
+                                              ? RouteEvalStatus::NoMoreRoutes
+                                              : RouteEvalStatus::HasMoreRoutes;
+            RouteMatchStatus match_status = cb(route_entry, eval_status);
+            if (match_status == RouteMatchStatus::Accept) {
+              return route_entry;
+            }
+            continue;
+          }
+
+          return route_entry;
```

Comment:
Besides, I suggest refactor the code like below to reduce nested if and redundant continue, which improves the readability. 
(from line 1914)

```
if (cb == nullptr) {
        return route_entry;
}
RouteEvalStatus eval_status = (std::next(route) == routes_.end())
                                  ? RouteEvalStatus::NoMoreRoutes
                                  : RouteEvalStatus::HasMoreRoutes;

RouteMatchStatus match_status = cb(route_entry, eval_status);
if (match_status == RouteMatchStatus::Accept) {
  return route_entry;
}
if (match_status == RouteMatchStatus::Continue &&
    eval_status == RouteEvalStatus::NoMoreRoutes) {
  return nullptr;
}

```

Reply:
Done. Is the current code good for you? btw, I'm not very familiar with c++, just wonder if it will be better to change `getRouteFromRoutes` to an inline function in this case?